### PR TITLE
PERF-#4851: Compute `dtypes` for binary operations that can only return bool type and the right operand is not a Modin object

### LINKED ISSUE #4851: PERF: compute `dtypes` for binary operations that can only return bool type

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4851 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

File: modin/core/dataframe/algebra/default2pandas/default.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def call(cls, func, obj_type=pandas.DataFrame, inplace=None, fn_name=None):
         if type(fn) == property:
             fn = cls.build_property_wrapper(fn)
 
-        def applyier(df, *args, **kwargs):
+        def applyier(df, *args, dtypes=None, **kwargs):
```
Comment (vnlitvinov) [inline modin/core/dataframe/algebra/default2pandas/default.py:71]: this is not used here, should we add an assert that it's always `None`?

**Review by vnlitvinov — COMMENTED** [review]
LGTM with one comment!

File: modin/core/dataframe/algebra/default2pandas/default.py (hunk 1 above)
Comment (anmyachev) [reply modin/core/dataframe/algebra/default2pandas/default.py:71]: There is a generic code that sets this argument even for the pure pandas case, so it's not always `None`.

**Review by RehanSD — COMMENTED** [review]
I did some tracing, and it seems that the dtypes argument is largely ignored by these codepaths? It only seems to be used when broadcast is set to True for the algebra operator, but none of these functions (le, gt,...) set broadcast to True, and it defaults to False. Additionally, shouldn't we add dtypes as a named, necessary kwarg for base's binary_op as well?

**Comment by anmyachev** [comment]
> I did some tracing, and it seems that the dtypes argument is largely ignored by these codepaths? It only seems to be used when broadcast is set to True for the algebra operator, but none of these functions (le, gt,...) set broadcast to True, and it defaults to False. Additionally, shouldn't we add dtypes as a named, necessary kwarg for base's binary_op as well?

Well, the changes now mostly cover cases where the right operand is not a Modin object, like just comparing with a scalar. https://github.com/modin-project/modin/blob/b0465fa4425247e7949881e93866c6b85bb10a1d/modin/core/dataframe/algebra/binary.py#L104

Yes, we should, I already opened the issue for that: https://github.com/modin-project/modin/issues/4856

**Comment by anmyachev** [comment]
@RehanSD @vnlitvinov ready to merge if there is no concerns

File: modin/core/dataframe/algebra/default2pandas/default.py (hunk 1 above)
Comment (YarShev) [reply modin/core/dataframe/algebra/default2pandas/default.py:71]: Is this dtypes used anywhere at all? Can you give an example?

Comment (anmyachev) [reply modin/core/dataframe/algebra/default2pandas/default.py:71]: It's used for example here: `df > 6` (code path will be through `map` function)

**Comment by vnlitvinov** [comment]
> Well, the changes now mostly cover cases where the right operand is not a Modin object, like just comparing with a scalar.

This bit of the information seems very important to understand what's going on.
I would rather you put it at least to the opening message of this PR.

File: modin/core/dataframe/algebra/default2pandas/default.py [hunk 2]
```diff
@@ -76,6 +76,8 @@ def applyier(df, *args, **kwargs):
             function under it and processes result so it is possible to create a valid
             query compiler from it.
             """
+            # pandas default implementation does not use argument
```
Comment (vnlitvinov) [inline modin/core/dataframe/algebra/default2pandas/default.py:79]:
```suggestion
            # pandas default implementation doesn't know how to handle `dtypes` keyword argument
```

**Review by YarShev — APPROVED** [review]

**Pull request merged**